pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/TableRebalancer.java

```diff
@@ -212,39 +218,132 @@ private IdealState getNextState(IdealState currentState, IdealState targetState,
+      }
+      LOGGER.debug(sb.toString());
+    }
 
-    if (srcMap == null) {
+    if (currentIdealStateSegmentHosts == null) {
       //segment can be missing if retention manager has deleted it
       LOGGER.info("Segment " + segmentId + " missing from current idealState. Skipping it.");
       return;
     }
 
     if (rebalanceUserConfig
+        // in downtime mode, set the current ideal state to target at one go
         .getBoolean(RebalanceUserConfigConstants.DOWNTIME, RebalanceUserConfigConstants.DEFAULT_DOWNTIME)) {
-      setTargetState(idealState, segmentId, targetMap);
+      LOGGER.debug("Downtime mode is enabled. Will set to target state at one go");
+      setTargetState(idealStateToUpdate, segmentId, targetIdealStateSegmentHosts);
+      ++_rebalancerStats.directTransitions;
       return;
     }
 
-    MapDifference difference = Maps.difference(targetMap, srcMap);
-    if (!difference.entriesInCommon().isEmpty()) {
-      // if there are entries in common, there won't be downtime
-      LOGGER.debug("Segment " + segmentId + " has common entries between current and expected ideal state");
-      setTargetState(idealState, segmentId, targetMap);
+    // we are in no-downtime mode
+    int minReplicasToKeepUp = rebalanceUserConfig
+        .getInt(RebalanceUserConfigConstants.MIN_REPLICAS_TO_KEEPUP_FOR_NODOWNTIME,
+            RebalanceUserConfigConstants.DEFAULT_MIN_REPLICAS_TO_KEEPUP_FOR_NODOWNTIME);
+    LOGGER.debug("No downtime mode is enabled. Need to keep {} serving replicas up while rebalancing",
+        minReplicasToKeepUp);
+
+    if (minReplicasToKeepUp >= currentIdealStateSegmentHosts.size()) {
+      // if the minimum number of serving replicas in rebalance config is
+      // greater than or equal to number of replicas of a segment, then it
+      // is impossible to honor the request. so we use the default number
+      // of minimum serving replicas we will keep for no downtime mode
+      // (currently 1)
+      minReplicasToKeepUp = RebalanceUserConfigConstants.DEFAULT_MIN_REPLICAS_TO_KEEPUP_FOR_NODOWNTIME;
+    }
+    MapDifference difference = Maps.difference(targetIdealStateSegmentHosts, currentIdealStateSegmentHosts);
+    if (difference.entriesInCommon().size() >= minReplicasToKeepUp) {
+      // if there are enough hosts in common between current and target ideal states
+      // to satisfy the min replicas condition, then there won't be any downtime
+      // and we can directly set the current ideal state to target ideal state
+      LOGGER.debug("Current and target ideal states have common hosts. Will set to target state at one go");
+      setTargetState(idealStateToUpdate, segmentId, targetIdealStateSegmentHosts);
+      ++_rebalancerStats.directTransitions;
     } else {
       // remove one entry
-      idealState.getInstanceStateMap(segmentId).remove(srcMap.keySet().stream().findFirst().get());
+      String hostToRemove = "";
```

Comment:
so the algorithm is to move one replica at a time until we reach a min number of replicas alive in target state, and then just move everything else. Would be good to note that in a comment above the method.

Reply:
+1

Reply:
updated javadoc